codex/slots/proofs/prover.nim

```diff
@@ -86,16 +87,18 @@ proc verify*(
   inputs: AnyProofInputs): Future[?!bool] {.async.} =
   ## Prove a statement using backend.
   ## Returns a future that resolves to a proof.
-
   self.backend.verify(proof, inputs)
 
 proc new*(
   _: type Prover,
   store: BlockStore,
-  backend: AnyBackend,
-  nSamples: int): Prover =
+  config: CodexConf): ?!Prover =
 
-  Prover(
-    backend: backend,
+  without backend =? initializeBackend(config), err:
```

Comment:
I'm really not a fan of this pattern, it violates the dependency injection principle, which has the nasty consequence of unnecessarily complicating testing. For example this usualy the main reasons for needing mocks.

It is perfectly fine to initialize the backend externally and then simply pass the required parameters to the constructor of the consumer class.

Reply:
I agree the hard call to `initializeBackend` is a problem. Putting all the details about the backend in codex.nim is a problem too. I'll try attaching this proc to a type and pass it to prover.new that way. We can have both the dependency injected and we can confine the knowledge about the proving system inside the slots/proofs files.

Reply:
I mean, just passing the backend is probably as good as it gets, I'm not sure why you'd want to pass in the factory instead, it doesn't seem to make anything "better"?